non-GPL/LibEbpfEvents/LibEbpfEvents.c

```diff
@@ -0,0 +1,92 @@
+// SPDX-License-Identifier: LicenseRef-Elastic-License-2.0
+
+/*
+ * Copyright 2021 Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License 2.0;
+ * you may not use this file except in compliance with the Elastic License 2.0.
+ */
+
+#define __aligned_u64 __u64 __attribute__((aligned(8)))
+#include "LibEbpfEvents.h"
+#include "Probe.skel.h"
+#include <bpf/libbpf.h>
+#include <errno.h>
+#include <stdio.h>
+
+struct ebpf_event_ctx {
+    struct ring_buffer *ringbuf;
+    struct Probe_bpf *probe;
+};
+
+/* This is just a thin wrapper that calls the event context's saved callback */
+static int ring_buf_cb(
+        void *ctx,
+        void *data,
+        size_t size)
+{
+    ebpf_event_handler_fn cb = ctx;
+    struct ebpf_event *evt = data;
+    return cb(evt, size);
+}
+
+int ebpf_event_ctx__new(
+        struct ebpf_event_ctx **ctx,
+        ebpf_event_handler_fn cb,
+        uint64_t features,
+        uint64_t events)
+{
+    int err;
+    *ctx = calloc(1, sizeof(struct ebpf_event_ctx));
+    if (ctx == NULL)
+        return -ENOMEM;
+
+    (*ctx)->probe = Probe_bpf__open();
+    if ((*ctx)->probe == NULL) {
+        free(*ctx);
+
+        /* Probe_bpf__open doesn't report errors, hard to find something that
+         * fits perfect here
+         */
+        return -ENOENT;
+    }
+
+    err = Probe_bpf__load((*ctx)->probe);
+    if (err != 0)
+        goto out_destroy_probe;
+
+    err = Probe_bpf__attach((*ctx)->probe);
+    if (err != 0)
+        goto out_destroy_probe;
+
+    struct ring_buffer_opts opts;
+    opts.sz = sizeof(opts);
+    (*ctx)->ringbuf = ring_buffer__new(
+            bpf_map__fd((*ctx)->probe->maps.ringbuf), ring_buf_cb, cb, &opts);
+
+    if ((*ctx)->ringbuf == NULL) {
+        /* ring_buffer__new doesn't report errors, hard to find something that
+         * fits perfect here
+         */
+        err = -ENOENT;
+        goto out_destroy_probe;
+    }
+
+    return ring_buffer__epoll_fd((*ctx)->ringbuf);
+
+out_destroy_probe:
+    Probe_bpf__destroy((*ctx)->probe);
+    free(*ctx);
+    return err;
+}
+
+int ebpf_event_ctx__next(
+        struct ebpf_event_ctx *ctx)
+{
+    int consumed = ring_buffer__poll(ctx->ringbuf, 0);
```

Comment:
Polling the ring buffer with no timeout leads to a busy CPU on my environment.


```
%CPU %MEM CMD
97.9  0.0 ./non-GPL/EventsTrace/EventsTrace
```

I'm not sure that keeping a full CPU busy is what we want.

```suggestion
    int consumed = ring_buffer__poll(ctx->ringbuf, 10);
```

Adding a default timeout `10ms` makes things better already:


```
%CPU %MEM CMD
 0.7  0.0 ./non-GPL/EventsTrace/EventsTrace
```


This is with `100ms`

```suggestion
    int consumed = ring_buffer__poll(ctx->ringbuf, 100);
```


```
%CPU %MEM CMD
 0.2  0.0 ./non-GPL/EventsTrace/EventsTrac
```


We can probably make this configurable as well.

Wdyt @rhysre ?

Reply:
Good call. I added an arg to `event_ctx__next`. I was looking at this from the perspective of non-blocking application that uses the epoll fd, but of course EventsTrace just blocks in a loop, so a timeout is needed.